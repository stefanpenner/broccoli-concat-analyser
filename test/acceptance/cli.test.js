'use strict';

const tmp = require('tmp');
const path = require('path');
const cp = require('child_process');
const fs = require('fs-extra');
const validate = require('html-validator');
const copyFixtures = require('../helpers/copy-fixtures');

const inputFixturePath = 'test/fixtures/input';

const inputFiles = ['1-test-app.js', '3-vendor.css', '8-test-support.css'];

const originalCwd = process.cwd();
let tmpPath;
let outPath;

function file(dir) {
  return fs.readFileSync(dir, 'utf-8');
}

function run() {
  let stdout = '';
  let stderr = '';

  return new Promise(resolve => {
    let ps = cp.spawn('node', [
      path.join(originalCwd, 'lib/cli.js'),
      './concat-stats-for'
    ], {
      cwd: tmpPath
    });

    ps.stdout.on('data', data => {
      stdout += data.toString();
    });

    ps.stderr.on('data', data => {
      stderr += data.toString();
    });

    ps.on('exit', code => {
      resolve({
        exitCode: code,
        stdout,
        stderr
      });
    });
  });
}

describe('CLI', function() {

  beforeEach(function() {
    tmpPath = fs.realpathSync(tmp.dirSync().name);
    outPath = path.join(tmpPath, 'concat-stats-for');
    fs.ensureDirSync(outPath);
    copyFixtures(inputFiles, inputFixturePath, outPath);
  });

  it('generates expected output', function() {
    return run()
      .then(result => {
        let exitCode = result.exitCode;
        let stdout = result.stdout;

        expect(exitCode).toEqual(0);
        expect(stdout).toContain(`visit file://${outPath}/index.html`);

        // write .out.json files
        inputFiles.forEach((inputFile) => {
          let outputFile = `${inputFile}.out.json`;
          expect(file(path.join(outPath, outputFile))).toMatchSnapshot();
        });

        // copies static output files
        expect(file(path.join(outPath, 'index.html'))).toBeTruthy();

        let data = fs.readFileSync(`${outPath}/index.html`, 'UTF8');
        expect(data).toContain('var SUMMARY = {');

        return validate({
          data,
          format: 'json'
        })
          .then(response => {
            let error = response.messages.find(msg => msg.type === 'error');
            if (error) {
              throw new Error(`HTML validation error in index.html at L${error.lastLine}:${error.firstColumn}-${error.lastColumn}: ${error.message}\n${error.extract}`);
            }

          })
      });
  });

});

