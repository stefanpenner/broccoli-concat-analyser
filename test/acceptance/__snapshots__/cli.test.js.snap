// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CLI generates expected output 1`] = `
"{
  \\"outputFile\\": \\"/assets/test-app.js\\",
  \\"files\\": [
    {
      \\"relativePath\\": \\"test-app/app.js\\",
      \\"sizes\\": {
        \\"raw\\": 624,
        \\"uglified\\": 377,
        \\"compressed\\": 120.451057165231
      }
    },
    {
      \\"relativePath\\": \\"test-app/initializers/app-version.js\\",
      \\"sizes\\": {
        \\"raw\\": 524,
        \\"uglified\\": 313,
        \\"compressed\\": 100.00313234142521
      }
    },
    {
      \\"relativePath\\": \\"test-app/initializers/container-debug-adapter.js\\",
      \\"sizes\\": {
        \\"raw\\": 606,
        \\"uglified\\": 430,
        \\"compressed\\": 137.38449490994518
      }
    },
    {
      \\"relativePath\\": \\"test-app/initializers/export-application-global.js\\",
      \\"sizes\\": {
        \\"raw\\": 1426,
        \\"uglified\\": 686,
        \\"compressed\\": 219.17619420516837
      }
    },
    {
      \\"relativePath\\": \\"test-app/resolver.js\\",
      \\"sizes\\": {
        \\"raw\\": 238,
        \\"uglified\\": 156,
        \\"compressed\\": 49.841816758026624
      }
    },
    {
      \\"relativePath\\": \\"test-app/router.js\\",
      \\"sizes\\": {
        \\"raw\\": 442,
        \\"uglified\\": 268,
        \\"compressed\\": 85.62568519968676
      }
    },
    {
      \\"relativePath\\": \\"test-app/templates/application.js\\",
      \\"sizes\\": {
        \\"raw\\": 406,
        \\"uglified\\": 324,
        \\"compressed\\": 103.51761942051684
      }
    }
  ]
}"
`;

exports[`CLI generates expected output 2`] = `
"{
  \\"outputFile\\": \\"/assets/vendor.css\\",
  \\"files\\": []
}"
`;

exports[`CLI generates expected output 3`] = `
"{
  \\"outputFile\\": \\"/assets/test-support.css\\",
  \\"files\\": [
    {
      \\"relativePath\\": \\"vendor/qunit/qunit.css\\",
      \\"sizes\\": {
        \\"raw\\": 8311,
        \\"compressed\\": 2146.4629921259843
      }
    },
    {
      \\"relativePath\\": \\"vendor/ember-qunit/test-container-styles.css\\",
      \\"sizes\\": {
        \\"raw\\": 578,
        \\"compressed\\": 149.2787401574803
      }
    }
  ]
}"
`;
